### PR TITLE
Oncoprinter

### PR DESCRIPTION
**Pull request by adamabeshouse**
Later PR will include MAF style input and hotspot annotation, and unifying data parsing between mutation mapper and oncoprinter

### CODE REVIEW

**Comment by adamabeshouse** [comment]
@onursumer @alisman can we review this? wanna merge and keep going

File: src/pages/staticPages/tools/oncoprinter/Oncoprinter.tsx [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        super(props);
+
+        (window as any).oncoprinter = this;
+
+        const self = this;
+
+        this.controlsHandlers = this.buildControlsHandlers();
+
+        this.controlsState = observable({
+            get showUnalteredColumns() {
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/Oncoprinter.tsx]: this are automatically made computed by calling observable on parent?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+    @action public setInput(data:string, genes:string, samples:string) {
+        this.setDataInput(data);
+        this.setGeneOrder(genes);
+        this.setSampleIdOrder(samples);
+    }
+
+    public readonly parsedInputLines = remoteData({
+        invoke:async()=>{
+            if (!this._dataInput) {
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: seems like a this data shaping is worthy of testing and should be decoupled from promise wrapper and fetching.   but if it's gonna be replaced by the unified data interface you told me about then lets discuss what makes sense

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+                        }), d=>genomeNexusKey2(d.annotation_summary.genomicLocation));
+                        // annotate lines in place - this keeps their place in parsedLines in original data order
+                        for (const line of toAnnotate) {
+                            const gnDatum = gnData[genomeNexusKey(line)];
+                            (line as OncoprinterInputLineType3).hugoGeneSymbol = gnDatum.annotation_summary.transcriptConsequenceSummary.hugoGeneSymbol;
+                            (line as OncoprinterInputLineType3).isHotspot = _.flatten(gnDatum.hotspots.annotation).filter(recurrentHotspotFilter).length > 0;
+                        }
+                    } catch (e) {
+                        // GenomeNexus error - filter out type3 lines because we don't know how to use them
+                        alert("Error fetching data from GenomeNexus service - throwing out Type 3 (MAF/mutation) input data rows.");
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: maybe good time to talk about how we can handle alerts

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    readonly hugoGeneSymbolToGene = remoteData({
+        await:()=>[this.hugoGeneSymbols],
+        invoke:async()=>{
+            const geneIds = this.hugoGeneSymbols.result!;
+            if (geneIds.length > 0) {
+                return _.keyBy(await client.fetchGenesUsingPOST({
+                    geneIdType: "HUGO_GENE_SYMBOL",
+                    geneIds
+                }), "hugoGeneSymbol");
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: is ts smart enough to check that the string "hugoGeneSymbol" is a key?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                    x=>(x.proteinPosStart !== undefined && x.proteinPosEnd !== undefined)
+                ) as Pick<Mutation, "entrezGeneId"|"proteinPosStart"|"proteinPosEnd">[];
+
+            const mutationPositionIdentifiers = _.values(countMutations(mutations));
+
+            const data = await client.fetchMutationCountsByPositionUsingPOST({
+                mutationPositionIdentifiers
+            });
+
+            return _.chain(data).groupBy(mutationCountByPositionKey).mapValues(counts=>_.sumBy(counts, "count")).value();
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: again, i don't think TS checks validity of the key string.  unless i'm wrong, lets use (o)=>o.count).

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts (hunk 4 above)
Comment (alisman) [reply src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: if not (didn't in my test) then lets use callback to get typesafety

File: src/pages/staticPages/tools/oncoprinter/OncoprinterTool.tsx [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        this.helpOpened = !this.helpOpened;
+    }
+
+    @autobind
+    private populateExampleData() {
+        this.dataInput = exampleData;
+    }
+
+    @autobind
+    private onDataInputChange(e: any) {
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterTool.tsx]: shall we just type these e's?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterTool.tsx [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        this.sampleOrderInput = e.target.value;
+    }
+
+    @action private doSubmit(dataInput:string) {
+        this.store.setInput(dataInput, this.geneOrderInput, this.sampleOrderInput);
+    }
+
+    @autobind
+    @action private onClickSubmit() {
+        if ($(`#${FILE_INPUT_ID}`).prop('files').length > 0) {
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterTool.tsx]: ref

File: src/pages/staticPages/tools/oncoprinter/Oncoprinter.tsx (hunk 1 above)
Comment (adamabeshouse) [reply src/pages/staticPages/tools/oncoprinter/Oncoprinter.tsx]: yeah https://mobx.js.org/refguide/object.html 
`Property getters will be automatically turned into derived properties, just like @computed would do.`

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts (hunk 2 above)
Comment (adamabeshouse) [reply src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: good point... probably worth doing now too, its not too much work

**Comment by adamabeshouse** [comment]
@cBioPortal/product this is ready for testing

**Comment by adamabeshouse** [comment]
oops sorry @jjgao.

@cBioPortal/product here it is now: http://cbioportal-frontend-pr-1706.herokuapp.com/oncoprinter

**Comment by jjgao** [comment]
Thanks, @adamabeshouse. It works great!

Are you going to support the same format as MutationMapper for this PR?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts (hunk 3 above)
Comment (adamabeshouse) [reply src/pages/staticPages/tools/oncoprinter/OncoprinterStore.ts]: lets talk about it for next round

**Pull request merged**